src/carousel/tp-carousel.ts

```diff
@@ -0,0 +1,413 @@
+		const slides: NodeListOf<TPCarouselSlideElement> | null | undefined = this.getSlideElements();
+
+		// Check if slides are available.
+		if ( slides ) {
+			// Tell the total number of slides.
+			return slides.length;
+		}
+
+		// Else return 0.
+		return 0;
+	}
+
+	/**
+	 * Set the current slide index.
+	 *
+	 * @param {number} index Slide index.
+	 */
+	setCurrentSlide( index: number ): void {
+		// Check if slide index is valid.
+		if ( index > this.getTotalSlides() || index <= 0 ) {
+			// Stop! It's not valid.
+			return;
+		}
+
+		// dispatch slide-set event.
+		this.dispatchEvent(
+			new CustomEvent( 'slide-set', {
+				bubbles: true,
+				detail: {
+					slideIndex: index,
+				},
+			} ),
+		);
+
+		// Set current slide index.
+		this.setAttribute( 'current-slide', index.toString() );
+	}
+
+	/**
+	 * Navigate to the next slide.
+	 */
+	next(): void {
+		// Flag that it is a programatic scroll.
+		this.flagProgramaticScroll();
+
+		// Initialize total slides variable.
+		const totalSlides: number = this.getTotalSlides();
+
+		// Check if we are at the last slide considering per view attribute.
+		if ( this.currentSlideIndex >= totalSlides - this.perView + 1 ) {
+			// Check if we are in infinite mode.
+			if ( 'yes' === this.getAttribute( 'infinite' ) ) {
+				// Yes, we are, and go back to first slide.
+				this.setCurrentSlide( 1 );
+			}
+
+			// Terminate.
+			return;
+		}
+
+		// Get next slide index by adding minimum of step or remaining number of slides.
+		const nextSlideIndex: number = this.currentSlideIndex + Math.min( this.step, totalSlides - this.currentSlideIndex - this.perView + 1 );
+
+		// Check if the next slide step is not taking it beyond the last slide.
+		if ( nextSlideIndex > ( totalSlides - this.perView + 1 ) ) {
+			// Yes, it is.
+			return;
+		}
+
+		// Everything is good, go to next slide.
+		this.setCurrentSlide( nextSlideIndex );
+	}
+
+	/**
+	 * Navigate to the previous slide.
+	 */
+	previous(): void {
+		// Flag that it is a programatic scroll.
+		this.flagProgramaticScroll();
+
+		// Check if we are at the first slide.
+		if ( this.currentSlideIndex <= 1 ) {
+			// Check if we are in infinite mode.
+			if ( 'yes' === this.getAttribute( 'infinite' ) ) {
+				this.setCurrentSlide( this.getTotalSlides() - this.perView + 1 );
+			}
+
+			// Terminate.
+			return;
+		}
+
+		// Get previous slide index.
+		const previousSlideNumber: number = this.currentSlideIndex;
+
+		// Check if the previous slide step is not taking it beyond the first slide.
+		if ( previousSlideNumber > this.step ) {
+			this.setCurrentSlide( previousSlideNumber - this.step );
+		} else {
+			this.setCurrentSlide( 1 );
+		}
+	}
+
+	/**
+	 * To set a flag if it is a programatic scroll.
+	 */
+	flagProgramaticScroll() {
+		// Set the flag to true.
+		this.isProgramaticScroll = true;
+
+		// Set the flag to false after 500ms.
+		setTimeout( () => {
+			// Set the flag to false.
+			this.isProgramaticScroll = false;
+		}, 500 );
```

Comment:
@harshdeep-gill Is this 500 ms depended on something? Could this be smaller or 0?